contracts/auto-withdrawer/src/contract.rs

```diff
@@ -106,107 +119,134 @@ fn execute_bond_with_ld_assets(
+        withdrawal_asset.amount += existing_bonding.withdrawal_amount;
+    }
     bondings_map().save(
         deps.storage,
-        &token_id,
+        &bonding_id,
         &BondingRecord {
             bonder: info.sender,
+            withdrawal_amount: withdrawal_asset.amount,
             deposit,
         },
     )?;
 
-    let withdrawal_voucher = WITHDRAWAL_VOUCHER_ADDRESS.load(deps.storage)?;
-    let msg = WasmMsg::Execute {
-        contract_addr: withdrawal_voucher.into_string(),
-        msg: to_json_binary(
-            &drop_staking_base::msg::withdrawal_voucher::ExecuteMsg::TransferNft {
-                recipient: env.contract.address.into_string(),
-                token_id,
-            },
-        )?,
-        funds: vec![],
-    };
-
     // TODO: attributes
-    Ok(Response::new().add_message(msg))
+    Ok(Response::new())
 }
 
 fn execute_unbond(
     deps: DepsMut<NeutronQuery>,
     info: MessageInfo,
-    token_id: String,
+    batch_id: Uint128,
 ) -> ContractResult<Response<NeutronMsg>> {
-    let bonding = bondings_map().load(deps.storage, &token_id)?;
+    let bonding_id = get_bonding_id(&info.sender, batch_id);
+    let bonding = bondings_map().load(deps.storage, &bonding_id)?;
     ensure_eq!(info.sender, bonding.bonder, ContractError::Unauthorized {});
-    bondings_map().remove(deps.storage, &token_id)?;
-
-    let withdrawal_voucher = WITHDRAWAL_VOUCHER_ADDRESS.load(deps.storage)?;
-
-    let nft_msg: CosmosMsg<NeutronMsg> = WasmMsg::Execute {
-        contract_addr: withdrawal_voucher.into_string(),
-        msg: to_json_binary(
-            &drop_staking_base::msg::withdrawal_voucher::ExecuteMsg::TransferNft {
-                recipient: info.sender.to_string(),
-                token_id,
-            },
-        )?,
-        funds: vec![],
-    }
-    .into();
-
-    let deposit_msg = BankMsg::Send {
-        to_address: info.sender.into_string(),
-        amount: bonding.deposit,
-    }
-    .into();
+    bondings_map().remove(deps.storage, &bonding_id)?;
+
+    let withdrawal_denom_prefix = WITHDRAWAL_DENOM_PREFIX.load(deps.storage)?;
+    let withdrawal_token_address = WITHDRAWAL_TOKEN_ADDRESS.load(deps.storage)?;
+    let withdrawal_denom =
+        get_full_withdrawal_denom(withdrawal_denom_prefix, withdrawal_token_address, batch_id);
+
+    let mut send_assets_vec = vec![Coin::new(
+        bonding.withdrawal_amount.u128(),
+        withdrawal_denom,
+    )];
+    send_assets_vec.extend(bonding.deposit);
+
+    let send_assets_msg: BankMsg = BankMsg::Send {
+        to_address: info.sender.to_string(),
+        amount: send_assets_vec,
+    };
 
     // TODO: attributes
-    Ok(Response::new().add_messages([nft_msg, deposit_msg]))
+    Ok(Response::new().add_messages([send_assets_msg]))
 }
 
 fn execute_withdraw(
     deps: DepsMut<NeutronQuery>,
     info: MessageInfo,
-    token_id: String,
+    batch_id: Uint128,
+    receiver: Option<Addr>,
+    amount: Uint128,
```

Comment:
Please elaborate why would anyone want to specify this amount? It is better to do everything in one transaction and earn the entire deposit anyway.

Reply:
It seems to me that the ability to withdraw only the entire amount is some kind of artificial limitation, why do we then introduce withdraw granularity in general? though, of course, there is no problem implementing such a limitation

Reply:
Granularity is implemented for users who wish to split their unbonding vouchers. But in this case, we are talking about an API which earns money for automatic bots which withdraw unbonded funds instead of the user and for the user, receive a deposit in process as a reward. I have no idea why would any bot ever would like to earn less than it might at the moment. I mean, bots have to pay for fees, too. And for this slightly complex feature, which would probably never be used, we have introduce some complex code, which might contain errors. I am really against this idea.